Add Alarm Control Panel

grahambrown11 opened the pull request:
# Implement Alarm Control Panel

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

A new `alarm_control_panel` that takes a list of `binary_sensors` that triggers the alarm when any of the sensor change to on, with the relevant API stuff to automatically create the Alarm Control Panel in HomeAssistant and provide state changes.

**Pull requests in progress:**
- ESPHome API: esphome/aioesphomeapi#427
- HomeAssistant: home-assistant/core#92357
- Docs: esphome/esphome-docs#2891

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
esphome:
  name: alarm-panel
  friendly_name: Alarm
  comment: Alarm Control Panel

esp8266:
  board: d1_mini

logger:

api:

alarm_control_panel:
  id: alarm
  name: Panel
  codes:
    - "1234"
  requires_code_to_arm: true
  arming_time: 10
  delay_time: 15
  trigger_time: 30
  binary_sensors:
    - input: zone_1
    - input: zone_2
      bypass_armed_home: true
    - input: ha_test
  on_state:
    then:
      - lambda: !lambda |-
          ESP_LOGD("TEST", "State change %s", id(alarm)->to_string(id(alarm)->get_state()).c_str());
  on_triggered:
    then:
      - switch.turn_on: siren
  on_cleared:
    then:
      - switch.turn_off: siren

binary_sensor:
  - platform: gpio
    id: zone_1
    name: Zone 1
    device_class: door
    pin:
      number: D1
      mode: INPUT_PULLUP
      inverted: True
  - platform: gpio
    id: zone_2
    name: Zone 2
    device_class: motion
    pin:
      number: D2
      mode: INPUT_PULLUP
      inverted: True
  - platform: homeassistant
    id: ha_test
    name: Zone 3
    entity_id: input_boolean.test_switch

switch:
  - platform: gpio
    id: siren
    name: Siren
    icon: mdi:alarm-bell
    pin: D7

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

grahambrown11 committed:
add alarm_control_panel owner

grahambrown11 committed:
add alarm_control_panel owner

grahambrown11 committed:
add alarm_control_panel owner

grahambrown11 committed:
Create alarm_control_panel Component

grahambrown11 committed:
Create alarm_control_panel Component

grahambrown11 committed:
Create alarm_control_panel Component

grahambrown11 committed:
lint fixes

grahambrown11 committed:
lint fixes

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
fix formatting

grahambrown11 committed:
fix formatting

grahambrown11 committed:
add alarm to test files

grahambrown11 committed:
add alarm to test files

grahambrown11 committed:
add alarm to webserver list entities

grahambrown11 committed:
add alarm to webserver list entities

grahambrown11 committed:
must be string

grahambrown11 committed:
must be string

grahambrown11 committed:
oops remove dsc_alarm_control_panel

grahambrown11 committed:
oops remove dsc_alarm_control_panel

grahambrown11 committed:
update ids to avoid conflics

grahambrown11 committed:
update ids to avoid conflics

dougiteixeira commented:
It would be very useful to have the option of configuring multiple codes to control the alarm, so it would be possible for each person to have their own personal code.

Is this viable?

grahambrown11 commented:
> It would be very useful to have the option of configuring multiple codes to control the alarm, so it would be possible for each person to have their own personal code.
> 
> Is this viable?

@dougiteixeira I'll add it to the list for future enhancements

nagyrobi commented:
> > It would be very useful to have the option of configuring multiple codes to control the alarm, so it would be possible for each person to have their own personal code.
> > Is this viable?
> 
> @dougiteixeira I'll add it to the list for future enhancements

Yes this is a must. Moreover, combined with a [Wiegand](https://esphome.io/components/wiegand.html) reader, not only numeric codes could be used, but also proxy cards, which are actually strings of various lengths. Each family member could have a separate proxy keyring or card.

grahambrown11 commented:
@nagyrobi 

> not only numeric codes could be used

The code is a string cos if your code started with a zero and it was stored as a numeric then the leading zero is dropped...

> combined with a Wiegand reader

For now you could use a HomeAssistant Automation to translate the many codes from the Wiegand reader
 into 1 to disarm... but will definatly add more codes in future release, for local ESPHome automation.
I just want to get this out as there are a lot of moving parts between ESPHome and HomeAssistant.

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
actions,triggers,multiple codes & multiple alarms

grahambrown11 committed:
code formatting

grahambrown11 committed:
another format

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 commented:
Multiple code support has now been added

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge remote-tracking branch 'upstream/dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
add log for checking code

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

jesserockz commented on esphome/components/alarm_control_panel/__init__.py:
```diff
@@ -0,0 +1,184 @@
+CONF_ON_TRIGGERED = "on_triggered"
+CONF_ON_CLEARED = "on_cleared"
+
+AlarmPanelNS = cg.esphome_ns.namespace("alarm_control_panel")
```
```suggestion
alarm_control_panel_ns = cg.esphome_ns.namespace("alarm_control_panel")
```

jesserockz commented on esphome/components/alarm_control_panel/__init__.py:
```diff
@@ -0,0 +1,184 @@
+CONF_ON_CLEARED = "on_cleared"
+
+AlarmPanelNS = cg.esphome_ns.namespace("alarm_control_panel")
+AlarmControlPanelComponent = AlarmPanelNS.class_("AlarmControlPanel", cg.Component)
```
Variable name should match the actual class name please.
```suggestion
AlarmControlPanel = AlarmPanelNS.class_("AlarmControlPanel", cg.Component)
```

jesserockz commented on esphome/components/alarm_control_panel/alarm_control_panel.h:
```diff
@@ -0,0 +1,200 @@
+  // the list of binary sensors that the alarm_panel monitors
+  std::vector<binary_sensor::BinarySensor *> sensors_;
+  // the list of sensor ids that the alarm_panel bypass when in armed_home
+  std::vector<uint32_t> bypass_when_home_;
```
You should really just store a pointer to the binary_sensor

grahambrown11 replied:
For the logic in the loop, I ignore the sensor when in the `bypass_when_home_` vector, I don't see how a pointer will help here... I have a vector of binary sensors above...

jesserockz commented on esphome/components/alarm_control_panel/__init__.py:
```diff
@@ -0,0 +1,184 @@
+    {
+        cv.GenerateID(): cv.declare_id(AlarmControlPanelComponent),
+        cv.Optional(CONF_CODES): cv.ensure_list(cv.string_strict),
+        cv.Optional(CONF_REQUIRES_CODE_TO_ARM, False): cv.boolean,
```
There should be a validation on `CONF_REQUIRES_CODE_TO_ARM` because it is ignored if you dont have any codes

jesserockz commented on esphome/components/alarm_control_panel/__init__.py:
```diff
@@ -0,0 +1,184 @@
+        cv.Optional(CONF_ARMING_AWAY_TIME, "0s"): cv.positive_time_period_seconds,
+        cv.Optional(CONF_DELAY_TIME, "0s"): cv.positive_time_period_seconds,
+        cv.Optional(CONF_TRIGGER_TIME, "0s"): cv.positive_time_period_seconds,
+        cv.Required(CONF_BINARY_SENSORS): cv.ensure_list(CONFIG_SCHEMA_BINARY_SENSOR),
```
I don't think binary_sensors should be **required**. personally I might want to write custom logic for my binary sensors (or any other sensor type) based on other factors and manually "trigger" the ACP to go into pending/triggered mode via yaml actions.

grahambrown11 replied:
I'm not sure I fully understand your usecase, in my mind an alarm system is a bunch of binary sensors that trigger the alarm when armed...
You can always use a template binary sensor to turn non binary sensors (like ultrasonic distance) into a binary on/off when over/under a certain value...

nagyrobi replied:
I think Jesse thinks of the possibility to use the system the other way around too: have actions triggered from the other parts of the config. Eg. an extreme case a binary sensor could change the state of the alarm fine-tuned from _Binary Sensor Filters_ / _Binary Sensor Automation_ without having to create copies of it with different settings for other uses (eg. the same binary sensor would be useful for both Alarm and A/C overriding, but with different trigger delay settings). 

So it would be very well feasible that someone wouldn't want to specify any `binary_sensors` in Alarm Control Panel, but would just use Alarm's Automation Actions to trigger it from various sensor's `on_...` sections.

I think this is just really about making binary_sensors optional.

jesserockz replied:
I'm not saying remove the `binary_sensors`, just don't force people to only have `binary_sensors` as the way to trigger the alarm from `armed` to `pending`. 
One use case could be instant triggering, where some sensors could be used to instantly trigger the alarm/sirens instead of going into pending/countdown mode.

grahambrown11 replied:
Cool, I guess a less opinionated system is more flexible for everyone

jesserockz commented on esphome/components/alarm_control_panel/__init__.py:
```diff
@@ -0,0 +1,184 @@
+        cv.Required(CONF_INPUT): cv.use_id(binary_sensor.BinarySensor),
+        cv.Optional(CONF_BYPASS_ARMED_HOME, default=False): cv.boolean,
+    }
+)
```
This should be wrapped in `cv.maybe_simple_id` to allow people not having to define a sub-object in yaml but just have a list of ids

jesserockz commented on esphome/components/alarm_control_panel/alarm_control_panel.cpp:
```diff
@@ -0,0 +1,268 @@
+      delay = this->arming_home_time_;
+    }
+    if ((millis() - this->last_update_) > delay) {
+      ESP_LOGD(TAG, "After Arming Time");
```
This log message seems a bit random

jesserockz commented on esphome/components/alarm_control_panel/alarm_control_panel.cpp:
```diff
@@ -0,0 +1,268 @@
+  // ARM_CUSTOM_BYPASS = 16
+  // ARM_VACATION = 32
+  return 1 + 2 + 8;
+}
```
This should be changed to use constants within here, and also should be set based on features users are actually using. For instance I would never use ARM_HOME in my house so I would not want HA to show it.

You don't have to implement all the modes in this PR, but at least ARM_HOME should be optional based on config.

jesserockz reviewed (changes requested):
One big thing missing from this is `calls` As per all of the other entities types that can be controlled (from with YAML or API) have a `Call` class that is used to `control` them. Please modify this to follow that same pattern.

I think the base ACP code should be more basic, and we should look at making platforms similar to all the other entity types with most of this code falling into the `template` platform under ACP.

There's probably more, but these requests might change the code too much to bother reviewing everything right now. Please reach out if you have questions.

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 commented:
> One big thing missing from this is `calls` As per all of the other entities types that can be controlled (from with YAML or API) have a `Call` class that is used to `control` them. Please modify this to follow that same pattern.

Sure I'll work on that, didn't know there were patterns to follow 😅 

> I think the base ACP code should be more basic, and we should look at making platforms similar to all the other entity types with most of this code falling into the `template` platform under ACP.

I don't understand - I thought the `template` platform was for using Lambda which is advanced 🤔 this is basic from a user of ESPHome is it not?

grahambrown11 committed:
PR review feedback

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
Review Feedback

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
remove commented out code

grahambrown11 committed:
clang-tidy & use pointers for bypass_when_home_

grahambrown11 committed:
more clang tidy

grahambrown11 committed:
hopefully the last clang tidy

grahambrown11 committed:
move is_armed_state

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
bump proto buf ids

grahambrown11 committed:
Implment platform concept

grahambrown11 committed:
Merge remote-tracking branch 'upstream/dev' into feature-alarm-panel

grahambrown11 committed:
after merge protobug regen

grahambrown11 committed:
add platform to acp test3

grahambrown11 committed:
Merge branch 'feature-alarm-panel' of github.com:grahambrown11/esphome into feature-alarm-panel

grahambrown11 committed:
apply clang format

grahambrown11 committed:
fix validation

grahambrown11 committed:
missed a platform

jesserockz committed:
Bitshift to make it easier to know the next flag value

jesserockz committed:
Dont allow arm home when not supported

jesserockz committed:
Allow code to compile without binary sensors
Use map to store flags for the mode of the sensors
Move simple setters to header
Rename delay time to pending time to match

jesserockz committed:
Aplit call and state into own files
Change to `publish_state`
Use LogString
Force implementation of virtual functions

jesserockz committed:
Fixes

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
fixes for conf changes & webserver state

grahambrown11 committed:
missed the logging

grahambrown11 committed:
fixes for ci

grahambrown11 committed:
oops changed the wrong time

grahambrown11 committed:
add trigger from HA

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

grahambrown11 committed:
back to LogString

RoganDawes commented:
I've been working on a UART integration with my Texecom Premier 832 alarm panel, which exposes the various sensors states as binary inputs, and provides controls for away arm/stay arm/disable, zone bypass, etc. The Texecom also has multiple partitions which can be armed and disarmed independently, but I'd love to be able to just extend the basic Alarm Control Panel classes to gain the automatic interfaces into e.g. the webserver and Home Assistant.

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel

jesserockz committed:
Use single underscore

grahambrown11 committed:
no underscore

jesserockz committed:
Add codeowner to template acp

jesserockz committed:
Remove extra log

jesserockz committed:
Allow actions and conditions to work without explicit id if only one in config

jesserockz committed:
Consistency

jesserockz committed:
Dont save pref by default, but give restore_mode config option

grahambrown11 committed:
Merge branch 'dev' into feature-alarm-panel
